src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+	std::transform(concurrent_chunks.begin(), concurrent_chunks.end(), concurrent_execution_ranges.begin(),
+	    [](const localized_chunk& chunk) { return chunk.execution_range; });
+
+	if(const auto overlapping_writes = detect_overlapping_writes(tsk, concurrent_execution_ranges); !overlapping_writes.empty()) {
+		auto error = fmt::format("{} has overlapping writes on N{} in", print_task_debug_label(tsk, true /* title case */), m_local_nid);
+		for(const auto& [bid, overlap] : overlapping_writes) {
+			fmt::format_to(std::back_inserter(error), " {} {}", print_buffer_debug_label(bid), overlap);
+		}
+		error += ". Choose a non-overlapping range mapper for this write access or constrain the split via experimental::constrain_split to make the access "
+		         "non-overlapping.";
+		utils::report_error(m_policy.overlapping_write_error, "{}", error);
+	}
+}
+
+void generator_impl::satisfy_task_buffer_requirements(batch& current_batch, const buffer_id bid, const task& tsk, const subrange<3>& local_execution_range,
+    const bool local_node_is_reduction_initializer, const std::vector<localized_chunk>& concurrent_chunks_after_split) //
+{
+	assert(!concurrent_chunks_after_split.empty());
+
+	auto& buffer = m_buffers.at(bid);
+
+	dense_map<memory_id, box_vector<3>> required_contiguous_allocations(m_memories.size());
+
+	box_vector<3> accessed_boxes; // which elements have are accessed (to figure out applying receives)
+	box_vector<3> consumed_boxes; // which elements are accessed with a consuming access (these need to be preserved across resizes)
+
+	const auto& bam = tsk.get_buffer_access_map();
+	for(const auto mode : bam.get_access_modes(bid)) {
+		const auto req = bam.get_mode_requirements(bid, mode, tsk.get_dimensions(), local_execution_range, tsk.get_global_size());
+		accessed_boxes.append(req.get_boxes());
+		if(access::mode_traits::is_consumer(mode)) { consumed_boxes.append(req.get_boxes()); }
+	}
+
+	// reductions can introduce buffer reads if they do not initialize_to_identity (but they cannot be split), so we evaluate them first
+	assert(std::count_if(tsk.get_reductions().begin(), tsk.get_reductions().end(), [=](const reduction_info& r) { return r.bid == bid; }) <= 1
+	       && "task defines multiple reductions on the same buffer");
+	const auto reduction = std::find_if(tsk.get_reductions().begin(), tsk.get_reductions().end(), [=](const reduction_info& r) { return r.bid == bid; });
+	if(reduction != tsk.get_reductions().end()) {
+		for(const auto& chunk : concurrent_chunks_after_split) {
+			required_contiguous_allocations[chunk.memory_id].push_back(scalar_reduction_box);
+		}
+		const auto include_current_value = local_node_is_reduction_initializer && reduction->init_from_buffer;
+		if(concurrent_chunks_after_split.size() > 1 || include_current_value) {
+			// we insert a host-side reduce-instruction in the multi-chunk scenario; its result will end up in the host buffer allocation
```

Comment:
Nope, we don't ever use SYCL's `include_current_buffer_value` because Acpp doesn't implement it. We treat single-chunk reductions that `initialize_to_identity` as ordinary writes in terms of Celerity buffer tracking because the reduction is complete once finished by SYCL. When we need to include the current value we insert a local host-side reduction that reduces the buffer contents and the output from the one (device) chunk.

Reply:
I see, maybe the comment could reflect that!

Reply:
Comment updated.